hail/src/main/scala/is/hail/expr/types/physical/PCanonicalBinary.scala

```diff
@@ -86,52 +79,54 @@ class PCanonicalBinary(val required: Boolean) extends PBinary {
 
   def constructAtAddress(mb: MethodBuilder, addr: Code[Long], region: Code[Region], srcPType: PType, srcAddress: Code[Long], forceDeep: Boolean): Code[Unit] = {
     val srcBinary = srcPType.asInstanceOf[PBinary]
-
-    val srcAddrVar = mb.newLocal[Long]
-    val len = mb.newLocal[Int]
-    Code(
-      srcAddrVar := srcAddress,
-      len := srcBinary.loadLength(srcAddrVar),
-      Region.storeAddress(addr, constructOrCopyWithLen(mb, region, srcBinary, srcAddrVar, len, forceDeep))
-    )
+    Region.storeAddress(addr, constructOrCopy(mb, region, srcBinary, srcAddress, forceDeep))
   }
 
-  private def constructOrCopyWithLen(mb: MethodBuilder, region: Code[Region], srcBinary: PBinary, srcAddress: LocalRef[Long], len: LocalRef[Int], forceDeep: Boolean): Code[Long] = {
+  private def constructOrCopy(mb: MethodBuilder, region: Code[Region], srcBinary: PBinary, srcAddress: Code[Long], forceDeep: Boolean): Code[Long] = {
     if (srcBinary == this) {
       if (forceDeep) {
+        val srcAddrVar = mb.newLocal[Long]
+        val len = mb.newLocal[Int]
         val newAddr = mb.newLocal[Long]
         Code(
+          srcAddrVar := srcAddress,
+          len := srcBinary.loadLength(srcAddrVar),
           newAddr := allocate(region, len),
-          Region.copyFrom(srcAddress, newAddr, contentByteSize(len)),
+          Region.copyFrom(srcAddrVar, newAddr, contentByteSize(len)),
           newAddr)
       } else
         srcAddress
     } else {
-
+      val srcAddrVar = mb.newLocal[Long]
+      val len = mb.newLocal[Int]
       val newAddr = mb.newLocal[Long]
       Code(
+        srcAddrVar := srcAddress,
+        len := srcBinary.loadLength(srcAddrVar),
         newAddr := allocate(region, len),
         storeLength(newAddr, len),
-        Region.copyFrom(srcAddress + srcBinary.lengthHeaderBytes, newAddr + lengthHeaderBytes, len.toL),
+        Region.copyFrom(srcAddrVar + srcBinary.lengthHeaderBytes, newAddr + lengthHeaderBytes, len.toL),
```

Comment:
Why not copy all bytes. I believe you brought this up yesterday, confirming: Lena can be different between source and dest I think. Could you provide an example?

Reply:
this is the generic interface, I think. There's no guarantee that the length and data are stored together.